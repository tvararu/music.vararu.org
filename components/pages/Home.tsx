import { useState } from "react";

const files = [
  "/static/music/Clams Casino/Rainforest/01 Natural.mp3",
  "/static/music/Clams Casino/Rainforest/02 Treetop.mp3",
  "/static/music/Clams Casino/Rainforest/03 Waterfalls.mp3",
  "/static/music/Clams Casino/Rainforest/04 Drowning.mp3",
  "/static/music/Clams Casino/Rainforest/05 Gorilla.mp3"
];

const Button = ({ children, onClick }) => (
  <button
    className="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded"
    onClick={onClick}
  >
    {children}
  </button>
);

const Player = () => {
  const [count, setCount] = useState(0);
  const currentSong = files[count];
  const skipToPreviousSong = () =>
    setCount((files.length + count - 1) % files.length);
  const skipToNextSong = () => setCount((count + 1) % files.length);

  return (
    <>
      <p>{count}</p>
      <p>Current song: {currentSong}</p>
      <audio controls src={currentSong}>
        Your browser does not support the audio element.
      </audio>
      <Button onClick={skipToPreviousSong}>Previous song</Button>
      <Button onClick={skipToNextSong}>Next song</Button>
    </>
  );
};

const Home = () => (
  <>
    <Player />
  </>
);

export default Home;
