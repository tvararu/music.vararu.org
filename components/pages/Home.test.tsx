import { cleanup, fireEvent, render, act } from "@testing-library/react";
import Home from "./Home";

const times = (n, fn) => Array.from(Array(n)).forEach(fn);

const currentTrack = {
  play: jest.fn(),
  pause: jest.fn()
};
global.Audio = jest.fn(() => currentTrack);

afterEach(cleanup);

let getByText;

beforeEach(() => {
  getByText = render(<Home />).getByText;
});

const playBtn = () => getByText("Play");
const pauseBtn = () => getByText("Pause");

const play = () => fireEvent.click(playBtn());
const pause = () => fireEvent.click(pauseBtn());
const next = () => fireEvent.click(getByText("Next"));
const previous = () => fireEvent.click(getByText("Previous"));

const firstSong = () => getByText(/Natural/);
const secondSong = () => getByText(/Treetop/);

const exists = el => expect(el).toBeInTheDocument();

const expectIsPlaying = () => {
  expect(currentTrack.play).toHaveBeenCalled();
  exists(pauseBtn());
};

const expectIsPaused = () => {
  expect(currentTrack.pause).toHaveBeenCalled();
  exists(playBtn());
};

it("allows playing and pausing", () => {
  play();
  expectIsPlaying();
  pause();
  expectIsPaused();
});

it("allows skipping to next track and previous track", () => {
  next();
  exists(secondSong());
  previous();
  exists(firstSong());
});

it("keeps playing after skipping", () => {
  play();
  next();
  expectIsPlaying();
});

it("loops to first song after last song and stops playing", () => {
  play();
  times(5, next);
  exists(firstSong());
  expectIsPaused();
});

it("stays at the first song when skipping backwards and stops playing", () => {
  play();
  previous();
  exists(firstSong());
  expectIsPaused();
});

describe("DOM events", () => {
  it("onplay plays", () => {
    act(() => currentTrack.onplay());
    expectIsPlaying();
  });

  it("onpause pauses", () => {
    play();
    act(() => currentTrack.onpause());
    expectIsPaused();
  });

  it("onended skips to next track", () => {
    play();
    act(() => currentTrack.onended());
    exists(secondSong());
    expectIsPlaying();
  });
});
