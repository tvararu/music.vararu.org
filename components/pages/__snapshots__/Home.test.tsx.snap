// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="jsx-1741780680 flex flex-col justify-center lg:h-screen"
  >
    <div
      class="jsx-1741780680 wrapper relative"
    >
      <div
        class="jsx-1741780680 flex flex-col items-center p-4"
      >
        <div
          class="jsx-1741780680 h-64 w-64 bg-gray-200 mb-4"
        />
        <h2
          class="jsx-1741780680 text-2xl font-bold"
        >
          01 Natural.mp3
        </h2>
      </div>
      <div
        class="jsx-1741780680 flex justify-center fixed lg:absolute bottom-0 w-screen"
      >
        <button
          class="bg-gray-200 hover:bg-gray-400 text-gray-900 font-bold py-2 px-4 text-3xl border-2 border-gray-900 m-3"
        >
          ⏮️
          <span
            class="jsx-140587568"
          >
            Previous
          </span>
        </button>
        <button
          class="bg-gray-200 hover:bg-gray-400 text-gray-900 font-bold py-2 px-4 text-3xl border-2 border-gray-900 m-3"
        >
          ▶️
          <span
            class="jsx-140587568"
          >
            Play
          </span>
        </button>
        <button
          class="bg-gray-200 hover:bg-gray-400 text-gray-900 font-bold py-2 px-4 text-3xl border-2 border-gray-900 m-3"
        >
          ⏭️
          <span
            class="jsx-140587568"
          >
            Next
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
